h2/src/main/org/h2/engine/Database.java

```diff
@@ -1339,38 +1339,38 @@ void close(boolean fromShutdownHook) {
     }
 
     private void closeImpl(boolean fromShutdownHook) {
-        try {
-            synchronized (this) {
-                if (closing) {
+        synchronized (this) {
+            if (closing) {
+                return;
+            }
+            closing = true;
+            stopServer();
+            if (!userSessions.isEmpty()) {
+                if (!fromShutdownHook) {
                     return;
                 }
-                closing = true;
-                stopServer();
+                trace.info("closing {0} from shutdown hook", databaseName);
+                closeAllSessionsExcept(null);
+            }
+            trace.info("closing {0}", databaseName);
+            if (eventListener != null) {
+                // allow the event listener to connect to the database
+                closing = false;
+                DatabaseEventListener e = eventListener;
+                // set it to null, to make sure it's called only once
+                eventListener = null;
+                e.closingDatabase();
                 if (!userSessions.isEmpty()) {
-                    if (!fromShutdownHook) {
-                        return;
-                    }
-                    trace.info("closing {0} from shutdown hook", databaseName);
-                    closeAllSessionsExcept(null);
-                }
-                trace.info("closing {0}", databaseName);
-                if (eventListener != null) {
-                    // allow the event listener to connect to the database
-                    closing = false;
-                    DatabaseEventListener e = eventListener;
-                    // set it to null, to make sure it's called only once
-                    eventListener = null;
-                    e.closingDatabase();
-                    if (!userSessions.isEmpty()) {
-                        // if a connection was opened, we can't close the database
-                        return;
-                    }
-                    closing = true;
-                }
-                if (!this.isReadOnly()) {
-                    removeOrphanedLobs();
+                    // if a connection was opened, we can't close the database
+                    return;
```

Comment:
I'd remove this piece - if you can't be bothered to obey the documented DbEventListener API, I don't see why we should leave the DB in some weird state.
Rather dump an exception and continue

Reply:
done